File: code/backend/php/tests/Unit/Core/Service/Auth/SessionServiceTest.php
<?php

declare(strict_types=1);

namespace Games\Paths\Tests\Unit\Core\Service\Auth;

use Games\Paths\Core\Domain\Auth\TokenInfo;
use Games\Paths\Core\Port\Auth\JwtPort;
use Games\Paths\Core\Port\Auth\TokenPersistencePort;
use Games\Paths\Core\Service\Auth\SessionService;
use PHPUnit\Framework\MockObject\MockObject;
use PHPUnit\Framework\TestCase;

class SessionServiceTest extends TestCase
{
    private JwtPort&MockObject $jwtPort;
    private TokenPersistencePort&MockObject $tokenPersistence;
    private SessionService $service;

    protected function setUp(): void
    {
        $this->jwtPort = $this->createMock(JwtPort::class);
        $this->tokenPersistence = $this->createMock(TokenPersistencePort::class);
        $this->service = new SessionService($this->jwtPort, $this->tokenPersistence, 5);
    }

    public function testRefreshSessionSuccess(): void
    {
        $oldToken = 'old-refresh-token';
        $userUuid = 'user-uuid-123';
        $username = 'testuser';
        
        $oldInfo = new TokenInfo($userUuid, $username, 'PLAYER', 'refresh', time(), time() + 3600, 'jti-1');
        
        $this->jwtPort->method('validateToken')->with($oldToken)->willReturn(true);
        $this->tokenPersistence->method('isRefreshTokenValid')->with($oldToken)->willReturn(true);
        $this->tokenPersistence->method('findUserIdByUuid')->with($userUuid)->willReturn(1);
        
        $this->tokenPersistence->expects($this->once())->method('revokeAllByUserUuid')->with($userUuid);
        
        $this->jwtPort->method('generateAccessToken')->willReturn('new-access-token');
        $this->jwtPort->method('generateRefreshToken')->willReturn('new-refresh-token');
        
        $newInfo = new TokenInfo($userUuid, $username, 'PLAYER', 'refresh', time(), time() + 7200, 'jti-2');
        $this->jwtPort->method('parseToken')->willReturnMap([
            [$oldToken, $oldInfo],
            ['new-refresh-token', $newInfo],
            ['new-access-token', new TokenInfo($userUuid, $username, 'PLAYER', 'access', time(), time() + 60, 'jti-3')]
        ]);

        $this->tokenPersistence->expects($this->once())->method('saveRefreshToken')->with($userUuid, 'new-refresh-token', $newInfo);
        $this->tokenPersistence->expects($this->once())->method('enforceTokenLimit')->with(1, 5);

        $result = $this->service->refreshSession($oldToken);

        $this->assertSame('new-access-token', $result->getAccessToken());
        $this->assertSame('new-refresh-token', $result->getRefreshToken());
        $this->assertSame($userUuid, $result->getUserUuid());
    }

    public function testRefreshSessionThrowsWhenInvalidToken(): void
    {
        $this->jwtPort->method('validateToken')->willReturn(false);
        
        $this->expectException(\RuntimeException::class);
        $this->expectExceptionMessage('INVALID_REFRESH_TOKEN');
        $this->expectExceptionCode(401);

        $this->service->refreshSession('invalid');
    }

    public function testRefreshSessionThrowsWhenNotRefreshToken(): void
    {
        $token = 'access-token-wrongly-used';
        $info = new TokenInfo('u', 'n', 'r', 'access', time(), time()+3600);
        
        $this->jwtPort->method('validateToken')->willReturn(true);
        $this->jwtPort->method('parseToken')->willReturn($info);
        
        $this->expectException(\RuntimeException::class);
        $this->expectExceptionMessage('NOT_A_REFRESH_TOKEN');
        
        $this->service->refreshSession($token);
    }

    public function testRefreshSessionThrowsWhenRevoked(): void
    {
        $token = 'revoked-token';
        $info = new TokenInfo('u', 'n', 'r', 'refresh', time(), time()+3600);
        
        $this->jwtPort->method('validateToken')->willReturn(true);
        $this->jwtPort->method('parseToken')->willReturn($info);
        $this->tokenPersistence->method('isRefreshTokenValid')->willReturn(false);
        
        $this->expectException(\RuntimeException::class);
        $this->expectExceptionMessage('REVOKED_REFRESH_TOKEN');
        
        $this->service->refreshSession($token);
    }

    public function testRefreshSessionThrowsWhenUserNotFound(): void
    {
        $token = 'token-for-missing-user';
        $info = new TokenInfo('u', 'n', 'r', 'refresh', time(), time()+3600);
        
        $this->jwtPort->method('validateToken')->willReturn(true);
        $this->jwtPort->method('parseToken')->willReturn($info);
        $this->tokenPersistence->method('isRefreshTokenValid')->willReturn(true);
        $this->tokenPersistence->method('findUserIdByUuid')->willReturn(null);
        
        $this->expectException(\RuntimeException::class);
        $this->expectExceptionMessage('USER_NOT_FOUND');
        
        $this->service->refreshSession($token);
    }

    public function testLogoutRevokesSpecificToken(): void
    {
        $token = 'logout-this';
        $this->tokenPersistence->expects($this->once())->method('revokeToken')->with($token);
        
        $this->service->logout($token);
    }

    public function testLogoutAllRevokesAllForUser(): void
    {
        $userUuid = 'user-1';
        $this->tokenPersistence->expects($this->once())->method('revokeAllByUserUuid')->with($userUuid);
        
        $this->service->logoutAll($userUuid);
    }

    public function testValidateAccessTokenSuccess(): void
    {
        $token = 'valid-access';
        $info = new TokenInfo('u', 'n', 'PLAYER', 'access', time(), time()+60);
        
        $this->jwtPort->method('validateToken')->with($token)->willReturn(true);
        $this->jwtPort->method('parseToken')->with($token)->willReturn($info);
        
        $result = $this->service->validateAccessToken($token);
        
        $this->assertSame($info, $result);
    }

    public function testValidateAccessTokenThrowsWhenInvalid(): void
    {
        $this->jwtPort->method('validateToken')->willReturn(false);
        
        $this->expectException(\RuntimeException::class);
        $this->expectExceptionMessage('INVALID_ACCESS_TOKEN');
        
        $this->service->validateAccessToken('bad');
    }

    public function testValidateAccessTokenThrowsWhenNotAccessToken(): void
    {
        $token = 'refresh-token-wrongly-used';
        $info = new TokenInfo('u', 'n', 'r', 'refresh', time(), time()+3600);
        
        $this->jwtPort->method('validateToken')->willReturn(true);
        $this->jwtPort->method('parseToken')->willReturn($info);
        
        $this->expectException(\RuntimeException::class);
        $this->expectExceptionMessage('NOT_AN_ACCESS_TOKEN');
        
        $this->service->validateAccessToken($token);
    }
}

File: code/backend/java/core/src/test/java/games/paths/core/service/auth/SessionServiceTest.java
package games.paths.core.service.auth;

import games.paths.core.model.auth.RefreshedSession;
import games.paths.core.model.auth.TokenInfo;
import games.paths.core.port.auth.JwtPort;
import games.paths.core.port.auth.TokenPersistencePort;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.HashMap;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.*;
import static org.mockito.Mockito.*;

/**
 * Unit tests for {@link SessionService}.
 * Covers all branches: refreshToken, logout, revokeAllSessions, validateAccessToken.
 */
@ExtendWith(MockitoExtension.class)
class SessionServiceTest {

    @Mock
    private JwtPort jwtPort;

    @Mock
    private TokenPersistencePort tokenPersistencePort;

    private SessionService sessionService;

    @BeforeEach
    void setUp() {
        sessionService = new SessionService(jwtPort, tokenPersistencePort, 5);
    }

    // === Constructor tests ===

    @Test
    @DisplayName("Should use default maxTokensPerUser via 2-arg constructor")
    void constructor_defaultMaxTokens() {
        SessionService svc = new SessionService(jwtPort, tokenPersistencePort);
        // Should not throw, uses default MAX_TOKENS_PER_USER = 5
        assertNotNull(svc);
    }

    // ==========================================================================
    // refreshToken
    // ==========================================================================

    @Nested
    @DisplayName("refreshToken Tests")
    class RefreshTokenTests {

        @Test
        @DisplayName("Should return null when refreshToken is null")
        void refreshToken_null() {
            assertNull(sessionService.refreshToken(null));
        }

        @Test
        @DisplayName("Should return null when refreshToken is blank")
        void refreshToken_blank() {
            assertNull(sessionService.refreshToken("  "));
        }

        @Test
        @DisplayName("Should return null when JWT validation fails")
        void refreshToken_invalidJwt() {
            when(jwtPort.validateToken("bad-token")).thenReturn(false);
            assertNull(sessionService.refreshToken("bad-token"));
        }

        @Test
        @DisplayName("Should return null when parseToken returns null")
        void refreshToken_parseReturnsNull() {
            when(jwtPort.validateToken("valid")).thenReturn(true);
            when(jwtPort.parseToken("valid")).thenReturn(null);
            assertNull(sessionService.refreshToken("valid"));
        }

        @Test
        @DisplayName("Should return null when token type is not 'refresh'")
        void refreshToken_notRefreshType() {
            when(jwtPort.validateToken("access-tok")).thenReturn(true);
            Map<String, Object> claims = new HashMap<>();
            claims.put("type", "access");
            when(jwtPort.parseToken("access-tok")).thenReturn(claims);
            assertNull(sessionService.refreshToken("access-tok"));
        }

        @Test
        @DisplayName("Should return null when token type is null")
        void refreshToken_nullType() {
            when(jwtPort.validateToken("no-type")).thenReturn(true);
            Map<String, Object> claims = new HashMap<>();
            claims.put("type", null);
            when(jwtPort.parseToken("no-type")).thenReturn(claims);
            assertNull(sessionService.refreshToken("no-type"));
        }

        @Test
        @DisplayName("Should return null when refresh token is revoked in DB")
        void refreshToken_revokedInDb() {
            when(jwtPort.validateToken("revoked-tok")).thenReturn(true);
            Map<String, Object> claims = new HashMap<>();
            claims.put("type", "refresh");
            when(jwtPort.parseToken("revoked-tok")).thenReturn(claims);
            when(tokenPersistencePort.isRefreshTokenValid("revoked-tok")).thenReturn(false);
            assertNull(sessionService.refreshToken("revoked-tok"));
        }

        @Test
        @DisplayName("Should return null when userData is null")
        void refreshToken_userDataNull() {
            when(jwtPort.validateToken("tok")).thenReturn(true);
            Map<String, Object> claims = new HashMap<>();
            claims.put("type", "refresh");
            when(jwtPort.parseToken("tok")).thenReturn(claims);
            when(tokenPersistencePort.isRefreshTokenValid("tok")).thenReturn(true);
            when(tokenPersistencePort.findUserByRefreshToken("tok")).thenReturn(null);
            assertNull(sessionService.refreshToken("tok"));
        }

        @Test
        @DisplayName("Should return null when userData is missing required uuid")
        void refreshToken_missingUuid() {
            when(jwtPort.validateToken("tok")).thenReturn(true);
            Map<String, Object> claims = new HashMap<>();
            claims.put("type", "refresh");
            when(jwtPort.parseToken("tok")).thenReturn(claims);
            when(tokenPersistencePort.isRefreshTokenValid("tok")).thenReturn(true);

            Map<String, Object> userData = new HashMap<>();
            userData.put("uuid", null);
            userData.put("username", "guest_1");
            userData.put("id", 42L);
            when(tokenPersistencePort.findUserByRefreshToken("tok")).thenReturn(userData);

            assertNull(sessionService.refreshToken("tok"));
        }

        @Test
        @DisplayName("Should return null when userData is missing username")
        void refreshToken_missingUsername() {
            when(jwtPort.validateToken("tok")).thenReturn(true);
            Map<String, Object> claims = new HashMap<>();
            claims.put("type", "refresh");
            when(jwtPort.parseToken("tok")).thenReturn(claims);
            when(tokenPersistencePort.isRefreshTokenValid("tok")).thenReturn(true);

            Map<String, Object> userData = new HashMap<>();
            userData.put("uuid", "u-1");
            userData.put("username", null);
            userData.put("id", 42L);
            when(tokenPersistencePort.findUserByRefreshToken("tok")).thenReturn(userData);

            assertNull(sessionService.refreshToken("tok"));
        }

        @Test
        @DisplayName("Should return null when userData is missing id")
        void refreshToken_missingId() {
            when(jwtPort.validateToken("tok")).thenReturn(true);
            Map<String, Object> claims = new HashMap<>();
            claims.put("type", "refresh");
            when(jwtPort.parseToken("tok")).thenReturn(claims);
            when(tokenPersistencePort.isRefreshTokenValid("tok")).thenReturn(true);

            Map<String, Object> userData = new HashMap<>();
            userData.put("uuid", "u-1");
            userData.put("username", "guest_1");
            // no id
            when(tokenPersistencePort.findUserByRefreshToken("tok")).thenReturn(userData);

            assertNull(sessionService.refreshToken("tok"));
        }

        @Test
        @DisplayName("Should successfully refresh token with non-null role")
        void refreshToken_successWithRole() {
            String oldRefresh = "valid-refresh";
            when(jwtPort.validateToken(oldRefresh)).thenReturn(true);

            Map<String, Object> claims = new HashMap<>();
            claims.put("type", "refresh");
            when(jwtPort.parseToken(oldRefresh)).thenReturn(claims);
            when(tokenPersistencePort.isRefreshTokenValid(oldRefresh)).thenReturn(true);

            Map<String, Object> userData = new HashMap<>();
            userData.put("uuid", "u-42");
            userData.put("username", "testuser");
            userData.put("role", "ADMIN");
            userData.put("id", 100L);
            when(tokenPersistencePort.findUserByRefreshToken(oldRefresh)).thenReturn(userData);

            when(jwtPort.generateAccessToken("u-42", "testuser", "ADMIN")).thenReturn("new-access");
            when(jwtPort.generateRefreshToken("u-42")).thenReturn("new-refresh");
            when(jwtPort.getRefreshTokenExpirationMs()).thenReturn(999999L);
            when(jwtPort.getAccessTokenExpirationMs()).thenReturn(500000L);

            RefreshedSession session = sessionService.refreshToken(oldRefresh);

            assertNotNull(session);
            assertAll(
                () -> assertEquals("u-42", session.getUserUuid()),
                () -> assertEquals("testuser", session.getUsername()),
                () -> assertEquals("ADMIN", session.getRole()),
                () -> assertEquals("new-access", session.getAccessToken()),
                () -> assertEquals("new-refresh", session.getRefreshToken()),
                () -> assertEquals(500000L, session.getAccessTokenExpiresAt()),
                () -> assertEquals(999999L, session.getRefreshTokenExpiresAt())
            );

            verify(tokenPersistencePort).revokeAllUserTokens(100L);
            verify(tokenPersistencePort).storeRefreshToken(eq(100L), eq("new-refresh"), anyString());
        }

        @Test
        @DisplayName("Should default role to PLAYER when role is null")
        void refreshToken_nullRoleDefaultsToPlayer() {
            String oldRefresh = "refresh-null-role";
            when(jwtPort.validateToken(oldRefresh)).thenReturn(true);

            Map<String, Object> claims = new HashMap<>();
            claims.put("type", "refresh");
            when(jwtPort.parseToken(oldRefresh)).thenReturn(claims);
            when(tokenPersistencePort.isRefreshTokenValid(oldRefresh)).thenReturn(true);

            Map<String, Object> userData = new HashMap<>();
            userData.put("uuid", "u-1");
            userData.put("username", "guest_u1");
            userData.put("role", null); // null role
            userData.put("id", 50L);
            when(tokenPersistencePort.findUserByRefreshToken(oldRefresh)).thenReturn(userData);

            when(jwtPort.generateAccessToken("u-1", "guest_u1", "PLAYER")).thenReturn("acc");
            when(jwtPort.generateRefreshToken("u-1")).thenReturn("ref");
            when(jwtPort.getRefreshTokenExpirationMs()).thenReturn(1000L);
            when(jwtPort.getAccessTokenExpirationMs()).thenReturn(500L);

            RefreshedSession session = sessionService.refreshToken(oldRefresh);

            assertNotNull(session);
            assertEquals("PLAYER", session.getRole());
            verify(jwtPort).generateAccessToken("u-1", "guest_u1", "PLAYER");
        }
    }

    // ==========================================================================
    // logout
    // ==========================================================================

    @Nested
    @DisplayName("logout Tests")
    class LogoutTests {

        @Test
        @DisplayName("Should return false when refreshToken is null")
        void logout_null() {
            assertFalse(sessionService.logout(null));
        }

        @Test
        @DisplayName("Should return false when refreshToken is blank")
        void logout_blank() {
            assertFalse(sessionService.logout("  "));
        }

        @Test
        @DisplayName("Should delegate to persistence and return true on success")
        void logout_success() {
            when(tokenPersistencePort.revokeRefreshToken("tok-1")).thenReturn(true);
            assertTrue(sessionService.logout("tok-1"));
            verify(tokenPersistencePort).revokeRefreshToken("tok-1");
        }

        @Test
        @DisplayName("Should return false when persistence reports failure")
        void logout_notFound() {
            when(tokenPersistencePort.revokeRefreshToken("unknown")).thenReturn(false);
            assertFalse(sessionService.logout("unknown"));
        }
    }

    // ==========================================================================
    // revokeAllSessions
    // ==========================================================================

    @Nested
    @DisplayName("revokeAllSessions Tests")
    class RevokeAllSessionsTests {

        @Test
        @DisplayName("Should return false when userUuid is null")
        void revokeAll_null() {
            assertFalse(sessionService.revokeAllSessions(null));
        }

        @Test
        @DisplayName("Should return false when userUuid is blank")
        void revokeAll_blank() {
            assertFalse(sessionService.revokeAllSessions("   "));
        }

        @Test
        @DisplayName("Should return false when user is not found (userId < 0)")
        void revokeAll_userNotFound() {
            when(tokenPersistencePort.findUserIdByUuid("unknown-uuid")).thenReturn(-1L);
            assertFalse(sessionService.revokeAllSessions("unknown-uuid"));
        }

        @Test
        @DisplayName("Should revoke all tokens and return true on success")
        void revokeAll_success() {
            when(tokenPersistencePort.findUserIdByUuid("u-1")).thenReturn(10L);
            when(tokenPersistencePort.revokeAllUserTokens(10L)).thenReturn(3);

            assertTrue(sessionService.revokeAllSessions("u-1"));
            verify(tokenPersistencePort).revokeAllUserTokens(10L);
        }
    }

    // ==========================================================================
    // validateAccessToken
    // ==========================================================================

    @Nested
    @DisplayName("validateAccessToken Tests")
    class ValidateAccessTokenTests {

        @Test
        @DisplayName("Should return null when accessToken is null")
        void validate_null() {
            assertNull(sessionService.validateAccessToken(null));
        }

        @Test
        @DisplayName("Should return null when accessToken is blank")
        void validate_blank() {
            assertNull(sessionService.validateAccessToken("  "));
        }

        @Test
        @DisplayName("Should return null when JWT validation fails")
        void validate_invalidJwt() {
            when(jwtPort.validateToken("bad")).thenReturn(false);
            assertNull(sessionService.validateAccessToken("bad"));
        }

        @Test
        @DisplayName("Should return null when parseToken returns null")
        void validate_parseNull() {
            when(jwtPort.validateToken("tok")).thenReturn(true);
            when(jwtPort.parseToken("tok")).thenReturn(null);
            assertNull(sessionService.validateAccessToken("tok"));
        }

        @Test
        @DisplayName("Should return null when token type is not 'access'")
        void validate_notAccessType() {
            when(jwtPort.validateToken("tok")).thenReturn(true);
            Map<String, Object> claims = new HashMap<>();
            claims.put("type", "refresh");
            when(jwtPort.parseToken("tok")).thenReturn(claims);
            assertNull(sessionService.validateAccessToken("tok"));
        }

        @Test
        @DisplayName("Should return null when type is null")
        void validate_nullType() {
            when(jwtPort.validateToken("tok")).thenReturn(true);
            Map<String, Object> claims = new HashMap<>();
            claims.put("type", null);
            when(jwtPort.parseToken("tok")).thenReturn(claims);
            assertNull(sessionService.validateAccessToken("tok"));
        }

        @Test
        @DisplayName("Should return null when subject (userUuid) is null")
        void validate_nullSubject() {
            when(jwtPort.validateToken("tok")).thenReturn(true);
            Map<String, Object> claims = new HashMap<>();
            claims.put("type", "access");
            claims.put("sub", null);
            when(jwtPort.parseToken("tok")).thenReturn(claims);
            assertNull(sessionService.validateAccessToken("tok"));
        }

        @Test
        @DisplayName("Should return valid TokenInfo on success with all claims")
        void validate_success() {
            when(jwtPort.validateToken("tok")).thenReturn(true);
            Map<String, Object> claims = new HashMap<>();
            claims.put("type", "access");
            claims.put("sub", "u-1");
            claims.put("username", "guest_u1");
            claims.put("role", "PLAYER");
            claims.put("jti", "jti-123");
            claims.put("iat", 1000L);
            claims.put("exp", 2000L);
            when(jwtPort.parseToken("tok")).thenReturn(claims);

            TokenInfo info = sessionService.validateAccessToken("tok");

            assertNotNull(info);
            assertAll(
                () -> assertEquals("u-1", info.getUserUuid()),
                () -> assertEquals("guest_u1", info.getUsername()),
                () -> assertEquals("PLAYER", info.getRole()),
                () -> assertEquals("access", info.getType()),
                () -> assertEquals("jti-123", info.getTokenId()),
                () -> assertEquals(1000L, info.getIssuedAt()),
                () -> assertEquals(2000L, info.getExpiresAt())
            );
        }

        @Test
        @DisplayName("Should handle iat/exp as Integer instead of Long")
        void validate_iatExpAsInteger() {
            when(jwtPort.validateToken("tok")).thenReturn(true);
            Map<String, Object> claims = new HashMap<>();
            claims.put("type", "access");
            claims.put("sub", "u-1");
            claims.put("iat", 1234);  // Integer, not Long
            claims.put("exp", 5678);
            when(jwtPort.parseToken("tok")).thenReturn(claims);

            TokenInfo info = sessionService.validateAccessToken("tok");

            assertNotNull(info);
            assertEquals(1234L, info.getIssuedAt());
            assertEquals(5678L, info.getExpiresAt());
        }

        @Test
        @DisplayName("Should handle iat/exp as null (non-Number)")
        void validate_iatExpNull() {
            when(jwtPort.validateToken("tok")).thenReturn(true);
            Map<String, Object> claims = new HashMap<>();
            claims.put("type", "access");
            claims.put("sub", "u-1");
            // iat and exp not in map — so they'll be null
            when(jwtPort.parseToken("tok")).thenReturn(claims);

            TokenInfo info = sessionService.validateAccessToken("tok");

            assertNotNull(info);
            assertEquals(0L, info.getIssuedAt());
            assertEquals(0L, info.getExpiresAt());
        }

        @Test
        @DisplayName("Should handle iat/exp as String (non-Number, should default to 0)")
        void validate_iatExpAsString() {
            when(jwtPort.validateToken("tok")).thenReturn(true);
            Map<String, Object> claims = new HashMap<>();
            claims.put("type", "access");
            claims.put("sub", "u-1");
            claims.put("iat", "not-a-number");
            claims.put("exp", "also-not");
            when(jwtPort.parseToken("tok")).thenReturn(claims);

            TokenInfo info = sessionService.validateAccessToken("tok");

            assertNotNull(info);
            assertEquals(0L, info.getIssuedAt());
            assertEquals(0L, info.getExpiresAt());
        }
    }
}
